### DISCUSSION TITLE
What does Field(...) really do?

### DISCUSSION THREAD

**Discussion by tobias-feil-plana** [pydantic/pydantic#8188] [Question] [answered]
(pydantic v2.5, python3.9)

In the docs, it is written that you mark required fields with `...`: 

https://docs.pydantic.dev/latest/concepts/models/#required-fields

But I just tried it out and the field behaves the exact same, whether I include `...` or just leave it out (pun intended).

So apparently, it is unnecessary to include it and the docs wrongly encourage you to do so. Or am I missing something?

![image](https://github.com/pydantic/pydantic/assets/142786001/607139bd-f47b-4967-b84f-2476423bef71)

**Comment by Viicos** [comment]
I think `...` can be used when you want to _explicitly_ state a value is required. with `foo: str = Field(description="Hello")`, we could wrongfully think that the field is not required, because we have a ` = smth`. Having `foo: str = Field(..., description="Hello")`emphasize on the fact that it is indeed required.

`foo: str = ...` and `foo: str = Field(...)` should be avoided, it is redundant and you should use `foo: str` instead.

I suggest also taking a look at `typing.Annotated`, which allows defining required fields with metadata more intuitively:

`foo: Annotated[str, Field(description="Hello")]` (--> no use of ` = smth`, which could be misleading).

**Comment by samuelcolvin** [comment, marked as answer]
Field is mostly there to configure all the other stuff you can set via it's kwargs, e.g. it's currently the only way to set `alias` or `default_factory`.

In time I think we'll allow more and more stuff to be configured via specific annotations, similar to annotated-types.

@sydney-runkle if the docs are really saying it's a way mark a field as required, we should change that.

**Reply by Viicos** [reply]
@samuelcolvin  I think OP was asking specifically about `...` (see my answer above)

**Reply by samuelcolvin** [reply]
I think you're right, sorry.

One reason to use it might be that dataclass transforms/static type checking needs to see the `default=...` to understand that the field is required, I think I saw that.

**Reply by tobias-feil-plana** [reply]
Thank you for the answer! Do you mean in a scenario like this:

![image](https://github.com/pydantic/pydantic/assets/142786001/e5a436ca-4faf-4824-948c-95ddab9536e5)

Maybe Pycharm is special but for me, there's no difference w.r.t. type checking, whether I include it or not.

**Reply by Viicos** [reply]
I see no mention of the `Ellipsis`/`...` in the dataclass_transform PEP, but maybe it was required at some point by type checkers when implementation was WIP. pyright and mypy now fully support `dataclass_transform`, so `...` should be irrelevant from a typing perspective (but Pydantic still makes use of it). Imho it should be avoided in most cases.

For example, see this [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFACYCGMlAxgDaUDOTA%2BjCJSk8LhAFD9yAU2BRgSYfXIAKAFSUQaJmTlyA1gHdFygJRQAtAD5CxAFz8oVqAmZNB-AAJUaDO%2B07deICDIlTyViYEYVokfxAmAF4-SWkSXV1%2BNxYoAFlhGhl4EN0Laxs7BxSmKAAhZmE0sBF6GQhMukYWKIyaPMtrWxZi5tKCGQqmKpqpDoLKM0wUfCjxONkRYEoAV3oYKIA6baSCgCMp1Fmobc3BAd0gA). Pyright thinks `a` and `b` is not required, doesn't special case the default value assignment of `b`, and doesn't raise an error when instantiating `A` without any arguments

**Reply by samuelcolvin** [reply]
Here's the case where you need the `default` kwarg for both mypy and pyright, but I agree, I don't see a case where the ellipsis is required

```py
from pydantic import BaseModel, Field


class Okay(BaseModel):
    x: int = Field(default=4, title='X')


Okay()


class Wrong(BaseModel):
    x: int = Field(4, title='X')


Wrong()  # gives an error with mypy and pyright about x being required.
```

**Reply by samuelcolvin** [reply]
we'll update the docs in #8224.

**Reply by Viicos** [reply]
Nice!

Regarding your `Wrong` example, the `dataclass_transform` PEP requires `default` to be a kw-only argument (that's why it is not understood by type checkers when using the positional variant). I went more in depth about it here: https://github.com/pydantic/pydantic/discussions/7379#discussioncomment-6951557
